File: tests/test_track.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright Contributors to the OpenTimelineIO project

#include "utils.h"

#include <opentimelineio/clip.h>
#include <opentimelineio/stack.h>
#include <opentimelineio/track.h>

#include <iostream>

using namespace OTIO_NS;

int
main(int argc, char** argv)
{
    Tests tests;

    tests.add_test(
        "test_find_children", [] {
        SerializableObject::Retainer<Clip> cl = new Clip();
        SerializableObject::Retainer<Track> tr = new Track();
        tr->append_child(cl);
        OTIO_NS::ErrorStatus err;
        auto result = tr->find_children<Clip>(&err);
        assertEqual(result.size(), 1);
        assertEqual(result[0].value, cl.value);
    });
    tests.add_test(
        "test_find_children_search_range", [] {
        const TimeRange range(RationalTime(0.0, 24.0), RationalTime(24.0, 24.0));
        SerializableObject::Retainer<Clip> cl0 = new Clip();
        cl0->set_source_range(range);
        SerializableObject::Retainer<Clip> cl1 = new Clip();
        cl1->set_source_range(range);
        SerializableObject::Retainer<Clip> cl2 = new Clip();
        cl2->set_source_range(range);
        SerializableObject::Retainer<Track> tr = new Track();
        tr->append_child(cl0);
        tr->append_child(cl1);
        tr->append_child(cl2);
        OTIO_NS::ErrorStatus err;
        auto result = tr->find_children<Clip>(
            &err,
            TimeRange(RationalTime(0.0, 24.0), RationalTime(24.0, 24.0)));
        assertEqual(result.size(), 1);
        assertEqual(result[0].value, cl0.value);
        result = tr->find_children<Clip>(
            &err,
            TimeRange(RationalTime(24.0, 24.0), RationalTime(24.0, 24.0)));
        assertEqual(result.size(), 1);
        assertEqual(result[0].value, cl1.value);
        result = tr->find_children<Clip>(
            &err,
            TimeRange(RationalTime(48.0, 24.0), RationalTime(24.0, 24.0)));
        assertEqual(result.size(), 1);
        assertEqual(result[0].value, cl2.value);
        result = tr->find_children<Clip>(
            &err,
            TimeRange(RationalTime(0.0, 24.0), RationalTime(48.0, 24.0)));
        assertEqual(result.size(), 2);
        assertEqual(result[0].value, cl0.value);
        assertEqual(result[1].value, cl1.value);
        result = tr->find_children<Clip>(
            &err,
            TimeRange(RationalTime(24.0, 24.0), RationalTime(48.0, 24.0)));
        assertEqual(result.size(), 2);
        assertEqual(result[0].value, cl1.value);
        assertEqual(result[1].value, cl2.value);
        result = tr->find_children<Clip>(
            &err,
            TimeRange(RationalTime(0.0, 24.0), RationalTime(72.0, 24.0)));
        assertEqual(result.size(), 3);
        assertEqual(result[0].value, cl0.value);
        assertEqual(result[1].value, cl1.value);
        assertEqual(result[2].value, cl2.value);
    });
    tests.add_test(
        "test_find_children_shallow_search", [] {
        SerializableObject::Retainer<Clip> cl0 = new Clip();
        SerializableObject::Retainer<Clip> cl1 = new Clip();
        SerializableObject::Retainer<Stack> st = new Stack();
        st->append_child(cl1);
        SerializableObject::Retainer<Track> tr = new Track();
        tr->append_child(cl0);
        tr->append_child(st);
        OTIO_NS::ErrorStatus err;
        auto result = tr->find_children<Clip>(&err, std::nullopt, true);
        assertEqual(result.size(), 1);
        assertEqual(result[0].value, cl0.value);
        result = tr->find_children<Clip>(&err, std::nullopt, false);
        assertEqual(result.size(), 2);
        assertEqual(result[0].value, cl0.value);
        assertEqual(result[1].value, cl1.value);
    });

    tests.add_test(
        "test_find_children_stack", [] {

        SerializableObject::Retainer<Stack> stack = new Stack();
        SerializableObject::Retainer<Track> track = new Track;
        SerializableObject::Retainer<Clip>  clip  = new Clip;
        stack->append_child(track);
        track->append_child(clip);

        // Simple find.
        clip->set_source_range(
            TimeRange(RationalTime(0.0, 24.0), RationalTime(3.0, 24.0)));
        OTIO_NS::ErrorStatus err;
        auto items = stack->find_children(
            &err,
            TimeRange(RationalTime(0.0, 24.0), RationalTime(1.0, 24.0)));
        assertFalse(is_error(err));
        assertEqual(items.size(), 2);
        assertTrue(
            std::find(items.begin(), items.end(), track.value) != items.end());
        assertTrue(
            std::find(items.begin(), items.end(), clip.value) != items.end());

        // Set a short source range on the track.
        track->set_source_range(
            TimeRange(RationalTime(0.0, 24.0), RationalTime(2.0, 24.0)));
        items = stack->find_children(
            &err,
            TimeRange(RationalTime(2.0, 24.0), RationalTime(1.0, 24.0)));
        assertFalse(is_error(err));
        assertEqual(items.size(), 0);

        // Set a source range with a positive offset on the track.
        track->set_source_range(
            TimeRange(RationalTime(3.0, 24.0), RationalTime(3.0, 24.0)));
        items = stack->find_children(
            &err,
            TimeRange(RationalTime(2.0, 24.0), RationalTime(1.0, 24.0)));
        assertFalse(is_error(err));
        assertEqual(items.size(), 1);
        assertTrue(
            std::find(items.begin(), items.end(), track.value) != items.end());

        // Set a source range with a negative offset on the track.
        track->set_source_range(
            TimeRange(RationalTime(-1.0, 24.0), RationalTime(3.0, 24.0)));
        items = stack->find_children(
            &err,
            TimeRange(RationalTime(1.0, 24.0), RationalTime(1.0, 24.0)));
        assertFalse(is_error(err));
        assertEqual(items.size(), 2);
        assertTrue(
            std::find(items.begin(), items.end(), track.value) != items.end());
        assertTrue(
            std::find(items.begin(), items.end(), clip.value) != items.end());
    });

    tests.run(argc, argv);
    return 0;
}

File: tests/test_track.py
#!/usr/bin/env python
# SPDX-License-Identifier: Apache-2.0
# Copyright Contributors to the OpenTimelineIO project

import unittest

import opentimelineio as otio
import opentimelineio.test_utils as otio_test_utils


class TrackTests(unittest.TestCase, otio_test_utils.OTIOAssertions):

    def test_find_children(self):
        cl = otio.schema.Clip()
        tr = otio.schema.Track()
        tr.append(cl)
        result = tr.find_children(otio.schema.Clip)
        self.assertEqual(len(result), 1)
        self.assertEqual(result[0], cl)

    def test_find_children_search_range(self):
        range = otio.opentime.TimeRange(
            otio.opentime.RationalTime(0.0, 24.0),
            otio.opentime.RationalTime(24.0, 24.0))
        cl0 = otio.schema.Clip()
        cl0.source_range = range
        cl1 = otio.schema.Clip()
        cl1.source_range = range
        cl2 = otio.schema.Clip()
        cl2.source_range = range
        tr = otio.schema.Track()
        tr.append(cl0)
        tr.append(cl1)
        tr.append(cl2)
        result = tr.find_children(otio.schema.Clip, range)
        self.assertEqual(len(result), 1)
        self.assertEqual(result[0], cl0)

    def test_find_children_shallow_search(self):
        cl0 = otio.schema.Clip()
        cl1 = otio.schema.Clip()
        st = otio.schema.Stack()
        st.append(cl1)
        tr = otio.schema.Track()
        tr.append(cl0)
        tr.append(st)
        result = tr.find_children(otio.schema.Clip, shallow_search=True)
        self.assertEqual(len(result), 1)
        self.assertEqual(result[0], cl0)
        result = tr.find_children(otio.schema.Clip, shallow_search=False)
        self.assertEqual(len(result), 2)
        self.assertEqual(result[0], cl0)
        self.assertEqual(result[1], cl1)

    def test_find_children_stack(self):
        video_clip = otio.schema.Clip();
        video_clip.source_range = otio.opentime.TimeRange(
            otio.opentime.RationalTime(0.0, 30.0),
            otio.opentime.RationalTime(700.0, 30.0))
        audio_clip = otio.schema.Clip();
        audio_clip.source_range = otio.opentime.TimeRange(
            otio.opentime.RationalTime(0.0, 30.0),
            otio.opentime.RationalTime(704.0, 30.0))
        video_track = otio.schema.Track()
        audio_track = otio.schema.Track()
        stack = otio.schema.Stack()
        video_track.append(video_clip)
        audio_track.append(audio_clip)
        stack.append(video_track)
        stack.append(audio_track)

        time = otio.opentime.RationalTime(703.0, 30.0)
        one_frame = otio.opentime.RationalTime(1.0, 30.0)
        range = otio.opentime.TimeRange(time, one_frame)
        items = stack.find_children(search_range=range)
        self.assertEqual(len(items), 2)
        self.assertTrue(audio_clip in items)
        self.assertTrue(audio_track in items)

if __name__ == '__main__':
    unittest.main()
